go/felix/rules/dispatch.go

```diff
@@ -0,0 +1,217 @@
+	commonPrefix := stringutils.CommonPrefix(names)
+	log.WithField("commonPrefix", commonPrefix).Debug("Calculated common prefix")
+
+	// Then, divide the names into bins based on their next character.
+	prefixes := []string{}
+	prefixToNames := map[string][]string{}
+	for _, name := range names {
+		prefix := commonPrefix
+		if len(name) > len(commonPrefix) {
+			prefix = name[:len(commonPrefix)+1]
+		}
+		if _, present := prefixToNames[prefix]; !present {
+			// Record the prefixes in sorted order (if we iterate over the map, we get a
+			// random order, which we don't want).
+			prefixes = append(prefixes, prefix)
+		}
+		prefixToNames[prefix] = append(prefixToNames[prefix], name)
+	}
+
+	rootFromEndpointRules := make([]Rule, 0, len(prefixes)+2)
+	rootToEndpointRules := make([]Rule, 0, len(prefixes)+2)
+
+	// Now, iterate over the prefixes.  If there are multiple names in a prefix, we render a
+	// child chain for that prefix.  Otherwise, we render the rule directly to avoid the cost
+	// of an extra goto.
+	var chains []*Chain
+	for _, prefix := range prefixes {
+		ifaceNames := prefixToNames[prefix]
+		logCxt := log.WithFields(log.Fields{
+			"prefix":          prefix,
+			"namesWithPrefix": ifaceNames,
+		})
+		logCxt.Debug("Considering prefix")
+		if len(ifaceNames) > 1 {
+			// More than one name, render a prefix match in the root chain...
+			nextChar := prefix[len(commonPrefix):]
+			ifaceMatch := prefix + "+"
+			childFromChainName := dispatchFromEndpointChainName + "-" + nextChar
+			childToChainName := dispatchToEndpointChainName + "-" + nextChar
+			logCxt := logCxt.WithFields(log.Fields{
+				"childFromChainName": childFromChainName,
+				"childToChainName":   childToChainName,
+				"ifaceMatch":         ifaceMatch,
+			})
+			logCxt.Debug("Multiple interfaces with prefix, rendering child chain")
+			rootFromEndpointRules = append(rootFromEndpointRules, Rule{
+				Match: Match().InInterface(ifaceMatch),
+				Action: GotoAction{
+					Target: childFromChainName,
+				},
+			})
+			rootToEndpointRules = append(rootToEndpointRules, Rule{
+				Match: Match().OutInterface(ifaceMatch),
+				Action: GotoAction{
+					Target: childToChainName,
+				},
+			})
+
+			// ...and child chains.
+			childFromEndpointRules := make([]Rule, 0, len(ifaceNames)+2)
+			childToEndpointRules := make([]Rule, 0, len(ifaceNames)+2)
+			for _, name := range ifaceNames {
+				logCxt.WithField("ifaceName", name).Debug("Adding rule to child chains")
+				childFromEndpointRules = append(childFromEndpointRules, Rule{
+					Match: Match().InInterface(name),
+					Action: GotoAction{
+						Target: EndpointChainName(fromEndpointPfx, name),
+					},
+				})
+				childToEndpointRules = append(childToEndpointRules, Rule{
+					Match: Match().OutInterface(name),
+					Action: GotoAction{
+						Target: EndpointChainName(toEndpointPfx, name),
+					},
+				})
+			}
+			if dropAtEndOfChain {
+				logCxt.Debug("Adding drop rules at end of child chains.")
```

Comment:
I think worth commenting here that we use Goto instead of Jump, and drop at the end of child chains, so that a non-matching packet does not have to iterate through all of the chain tree.  (It took me a few minutes to see that.)